gmlc/networking/SocketFactory.h

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright (c) 2017-2022,
+Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance
+for Sustainable Energy, LLC.  See the top-level NOTICE for additional details.
+All rights reserved. SPDX-License-Identifier: BSD-3-Clause
+*/
+#pragma once
+
+#include "Socket.h"
+#include <asio/io_context.hpp>
+#include <asio/ip/tcp.hpp>
+
+#ifdef GMLC_NETWORKING_ENABLE_ENCRYPTION
+#include <asio/ssl.hpp>
+#endif
+
+#include <functional>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <system_error>
+#include <vector>
+
+namespace gmlc::networking {
+class SocketFactory : std::enable_shared_from_this<SocketFactory> {
```

Comment:
Would it be appropriate to make SocketFactory an abstract class here?

Reply:
Not necessarily in this PR, to be clear.

Reply:
What would the other SocketFactory classes be? An AsioSocketFactory, or split into AsioEncryptedSocketFactory and AsioUnencryptedSocketFactory?

Reply:
I think that will need some other bits of the Tcp* classes that use asio directly to be refactored. For the SocketFactory, maybe a preprocessor define could be used that switches the namespace of the socket created based on if boost::asio or standalone asio is used?

Reply:
I think there are various options that might work once we get around to it.